src/index.d.ts

```diff
@@ -0,0 +1,344 @@
+    /**
+     * the component for the route
+     */
+    Component?: React.ComponentType<any>
+    /**
+     * a method that returns the component for the route
+     */
+    getComponent?: (
+      props: RoutingState,
+    ) => React.ComponentType<any> | Promise<React.ComponentType<any>> | Promise<JSX.Element>
+    /**
+     * additional data for the route
+     */
+    data?: any
+    /**
+     * a method that returns additional data for the route
+     */
+    getData?: (args: GetDataArgs) => any
+    prepareParams?: (params: Map, match: Match) => Map
+    /**
+     *
+     * @returns never (RedirectException) | undefined | React.ReactElement<any> (typical)
+     */
+    render?: (args: RenderArgs) => never | undefined | React.ReactElement<any>
+    // Provide indexer allowing for any properties
+    [key: string]: any
+  }
+
+  interface RouteProps extends BaseRouteConfig {
+    children?: Array<React.ReactElement<RouteProps>> | React.ReactElement<RouteProps>
+  }
+
+  /**
+   * JSX Route
+   */
+  class Route extends React.Component<RouteProps> {}
+
+  /**
+   * e.g. Resolved JSX Route
+   */
+  interface RouteConfig extends BaseRouteConfig {
+    children?: RouteConfig[]
+  }
+
+  class HttpError {
+    public status: number
+    public data: any
+    constructor(status: number)
+  }
+
+  interface RedirectProps {
+    from: string
+    to: string
+  }
+
+  class Redirect extends React.Component<RedirectProps> {}
+
+  interface LinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
+    Component?: React.ComponentType<any>
+    to: string | LocationArg
+    // match: Match,  provided by withRouter
+    activeClassName?: string
+    activeStyle?: any
+    activePropName?: string
+    // router: Router, provided by withRouter
+    exact?: boolean
+    target?: string
+    childProps?: any
+  }
+
+  class Link extends React.Component<LinkProps> {
+    public onClick: (event: React.SyntheticEvent<any>) => void
+  }
+
+  interface WithRouter {
+    match: Match
+    router: Router
+  }
+
+  function withRouter<OriginalProps>(
+    Component: React.ComponentType<OriginalProps & WithRouter>,
+  ): React.ComponentType<OriginalProps>
+
+  class RedirectException {
+    constructor(location: string | LocationArg)
+  }
+
+  /**
+   * Create a route configuration from JSX configuration elements.
+   */
+  function makeRouteConfig(node: React.ReactNode): RouteConfig
+
+  // Improve these `any`s as needed
+  type BrowserRouter = any
+
+  interface CreateBrowserRouterArgs {
+    render?: (args: RenderArgs) => React.ReactElement<any>
+    renderPending?: (args: RenderArgs) => React.ReactElement<any>
+    renderReady?: (args: RenderReadyArgs) => React.ReactElement<any>
+    renderError?: (args: RenderArgs) => React.ReactElement<any>
+    [key: string]: any
+  }
+
+  function createBrowserRouter(options: CreateBrowserRouterArgs): BrowserRouter
+
+  // Improve these `any`s as needed
+  type FarceRouter = any
+
+  function createFarceRouter({
```

Comment:
another one here :) why not making a type out of these args and use that instead of `any`?

Reply:
As noted, this is a starting point.  Feel free to enhance.  I have typed everything I use (and more), now it's time for everyone to make it better.